### PR TITLE
Add `LogicalPlan::recompute_schema` for handling rewrite passes

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of https://github.com/apache/datafusion/issues/9637 

## Rationale for this change

`LogicalPlan::recompute_schema` is needed for several subtasks of https://github.com/apache/datafusion/issues/9637 (for example https://github.com/apache/datafusion/pull/10356 and https://github.com/apache/datafusion/pull/10405)

I felt making its own PR with tests might make those PRs easier to review

## What changes are included in this PR?

1. Add `LogicalPlan::recompute_schema`
2. Add tests for `LogicalPlan::recompute_schema`

## Are these changes tested?
Yes, new tests (though full coverage will come in  https://github.com/apache/datafusion/pull/10356 and https://github.com/apache/datafusion/pull/10405)

### CODE REVIEW

File: datafusion/expr/src/logical_plan/plan.rs [hunk 1]
```diff
@@ -467,6 +468,200 @@ impl LogicalPlan {
...
+    /// expressions. For example [`LogicalPlan::Filter`] schema is always the
+    /// same as its input schema.
+    ///
+    /// # Return value
+    /// Returns an error if there is some issue recomputing the schema.
+    ///
+    /// # Notes
+    ///
+    /// * Does not recursively recompute schema for input (child) plans.
+    pub fn recompute_schema(self) -> Result<Self> {
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/plan.rs:488]: All this logic is the same as `new_with_expr`, even some questionable code like for Filter and Union

File: datafusion/expr/src/logical_plan/plan.rs [hunk 2]
```diff
@@ -3138,4 +3350,86 @@ digraph {
         let actual = format!("{}", plan.display_indent());
         assert_eq!(expected.to_string(), actual)
     }
+
+    #[test]
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/plan.rs:3354]: Here are some basic tests that show what it does and how to use it.

File: datafusion/expr/src/logical_plan/plan.rs (hunk 1 above)
Comment (yyy1000) [reply datafusion/expr/src/logical_plan/plan.rs:488]: I have a question which may be stupid, what's the difference between this and getting the inputs and exprs of the LogicalPlan and pass them as params to `new_with_expr`? 🤔

Comment (alamb) [reply datafusion/expr/src/logical_plan/plan.rs:488]: The key difference is that to use `new_with_expr` it requires providing a copy (`Vec<Expr>` and `Vec<LogicalPlan>`) where as this function can be used after modifying the Exprs (or children) via methods such as `map_children` and `map_expressions`

Avoiding those copies is a key part of improve planning performance in PRs like https://github.com/apache/datafusion/pull/10356 (the change is basically to stop copying)

Comment (alamb) [reply datafusion/expr/src/logical_plan/plan.rs:493]: I pushed bd7b62f2aa76e33a4fcba7cffaf02920114192bb to try and clarify

**Comment by alamb** [comment]
PR https://github.com/apache/datafusion/pull/10443 to add comments based on @yyy1000 's feedback on this PR

**Pull request closed without merging**